### PR TITLE
Nodes: MaterialX Noise Functions Lib

### PR DESCRIPTION
**Pull request by sunag**
Related issue: https://github.com/mrdoob/three.js/issues/20541

**Description**

This is an early part of the implementation.

## NodeMaterial

- [x] Add `functional interface` to `GLSLNodeFunction`.
- [x] [Add support to array for function call](https://github.com/mrdoob/three.js/commit/87de46756cd95da34de7bbda7c2e7acdadd4aec8)
- [x] [Add includes for constructor argument](https://github.com/mrdoob/three.js/commit/0c3373fc71acafee4a8b2adca98a92cb6a6b5ca4)
- [x] [Convert to `nodeObjects` function call parameters and includes](https://github.com/mrdoob/three.js/commit/05208ebd5a06f18f83db38a8eb7a399220e884a7)

## ShaderNode

- [x] add `fn` method to create function using backend shader language. [ commit ](https://github.com/mrdoob/three.js/commit/7696f222bc75aeb6e02e092e63859fc0e51e3776)

## Example

- add `webgl_nodes_materialx_noise` [ commit ](https://github.com/mrdoob/three.js/commit/b6f2be1ab67f9a7c088f5af3a388746c72b024b1)

https://raw.githack.com/sunag/three.js/dev-materialx-functions-lib/examples/webgl_nodes_materialx_noise.html

image

*This contribution is funded by [Google via Igalia](https://igalia.com)*

### CODE REVIEW

File: examples/jsm/nodes/materialx/functions/lib/mx_hsv.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { fn } from '../../../Nodes.js';
+
+// Original shader code from:
+// https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/libraries/stdlib/genglsl/lib/mx_hsv.glsl
+
+export const mx_hsvtorgb = fn( `vec3 mx_hsvtorgb(vec3 hsv)
```
[the remark this answers is not in the document]
Comment (LeviPesin) [reply examples/jsm/nodes/materialx/functions/lib/mx_hsv.js:6]: Or maybe it should be another PR, after IfElseNode would be added.

[the remark this answers is not in the document]
Comment (sunag) [reply examples/jsm/nodes/materialx/functions/lib/mx_hsv.js:6]: > Or maybe it should be another PR, after IfElseNode would be added.

Yes, I think the same thing.

**Comment by mrdoob** [comment]
This is looking great!

**Comment by mrdoob** [comment]
Thanks!

**Comment by jstone-lucasfilm** [comment]
This looks really promising, thanks @sunag and @mrdoob.

Over time, I wonder if we might be able to increase the degree of _code sharing_ between the MaterialX and three.js codebases, so that future refinements to core shading code (e.g. noise functions such as `mx_worley_noise_vec3`) can be made in a central location and inherited by the other project when needed.

Based on work that was started at Autodesk, we have early JavaScript bindings for MaterialX libraries such as MaterialXGenGlsl, so one option would be to determine what's missing from those bindings that would be needed for three.js to call into this library directly, generating the required GLSL code when a MaterialX node (e.g. `worleynoise3d`) is referenced in an asset.

None of these considerations should block the great work that I'm seeing in progress, but it seems worthwhile to consider them for the future.

**Comment by mrdoob** [comment]
Thanks for the pointers!

> Over time, I wonder if we might be able to increase the degree of code sharing between the MaterialX and three.js codebases, so that future refinements to core shading code (e.g. noise functions such as mx_worley_noise_vec3) can be made in a central location and inherited by the other project when needed.

That would be great indeed! Maybe a npm package with all the snippets?

**Comment by jstone-lucasfilm** [comment]
@mrdoob That's a really good suggestion.  We currently package MaterialX-generated shader snippets for OSL in this way, allowing them to be individually accessed in tools that perform their own final shader generation.  It should be straightforward to extend this to GLSL in the future, and the results could indeed be shared directly through `npm` and other package managers.

For the formatting of this shader snippet library, would it be reasonable to have a separate GLSL file for each MaterialX node, including the definition of that node plus file references to any shared GLSL libraries upon which it depends?  That's the pattern we currently use for OSL, so it would likely be the most straightforward starting point for GLSL and other languages.

**Comment by mrdoob** [comment]
@sunag what do you think?

**Comment by sunag** [comment]
Thanksss @jstone-lucasfilm :) . Soon after reading the comment I thought of making a live example using MaterialX, and it worked perfectly with the GLSL code already in the MaterialX repository.

Live
https://raw.githack.com/sunag/three.js/dev-materialx-live/examples/webgl_nodes_materialx_live.html

Code
```js
const mxNoiseURL = "https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/libraries/stdlib/genglsl/lib/mx_noise.glsl";
const mxNoiseLive = await fetch( mxNoiseURL );
const mxNoise = code( await mxNoiseLive.text() );

const mxIncludes = [ mxNoise ];

const mx_perlin_noise_float = fn( 'float mx_perlin_noise_float( vec3 p )', mxIncludes );

//..

const material = new MeshPhysicalNodeMaterial();
material.colorNode = mx_perlin_noise_float( someCoordNode );
```

Using interfaces to identify the signature of the functions we can refine the code at will, including the dependencies.

In a way this is already possible without much effort, of course we can do something more organized creating a npm package if needed or create only one build in the MaterialX repo itself for this purpose, but I think the idea of @jstone-lucasfilm is very good.

**Comment by donmccurdy** [comment]
> For the formatting of this shader snippet library, would it be reasonable to have a separate GLSL file for each MaterialX node, including the definition of that node plus file references to any shared GLSL libraries upon which it depends? ...

In the npm ecosystem, the path an end-user imports is abtracted from the files on disk. Separate files for each node would be completely fine, there just needs to be something (usually a `src/index.js` entry) that defines a list of all entrypoints to the module:

```js
// index.js
export * from './nodes/nodeA.js';
export * from './nodes/nodeB.js';
export * from './nodes/nodeC.js';
// ...
```

Those files may in turn import from any shared GLSL libraries they require. The end-user only sees something like this:

```js
import { nodeB } from '@materialx/glsl';
```

Happy to help with the scaffolding if MaterialX wants to go in that direction!

**Comment by sunag** [comment]
@donmccurdy I made a [draft here](https://github.com/sunag/three.js/blob/620d79df16caf4a0feb44c8662c82f086aa4c48d/examples/jsm/nodes/materialx/MaterialXNodes.js) a few days ago but I think about doing a PR after the name of the nodes becomes more mature in the context of three.js. Anyway, I also think that would be the better way too.

**Comment by sunag** [comment]
@jstone-lucasfilm have you ever thought of putting an attribute of canvas position XY in `mtlx` this is just for the purpose of editing in a node editor? 

For example:
```xml
<texcoord name="texcoord1" type="vector2" posx="100" posy="50" />
```

This would represent a node in the editor at position `{ x:100, y:50 }`. I think it would allow exporting not only the material but the entire editing canvas, so it can be used in other node editors with the same arrangement.

**Comment by jstone-lucasfilm** [comment]
@sunag Although they're not yet leveraged in production tools, there are actually `xpos` and `ypos` attributes in MaterialX that address this use case:

https://materialx.org/assets/MaterialX.v1.38.Spec.pdf#page=42

We're planning to support this feature in the upcoming Open MaterialX Graph Editor, which is scheduled to appear on GitHub later this year:

https://materialx.org/assets/ASWF_OSD2022_MaterialX_OSL_Final.pdf#page=26

One caveat is that the `xpos` and `ypos` attributes are expressed as _multiples_ of the default width and height of a node in the application, rather than being literal pixel positions.  Our thinking is that this makes them more transferrable between applications with radically different node layouts, though we're open to new ideas from the community on improving this convention.

**Comment by sunag** [comment]
Ohh... Amazing!! Certainly `Open MaterialX Graph Editor` and `Three.js Node Editor` will have a great exchange soon.

**Comment by mrdoob** [comment]
@jstone-lucasfilm Any plans on generating wgsl files too? https://www.w3.org/TR/WGSL/

**Comment by jstone-lucasfilm** [comment]
@mrdoob Not yet, though if there's a good case to be made for direct generation of WGSL, I believe that would be a straightforward feature to add.  What are your thoughts on the benefits this would bring?

**Comment by sunag** [comment]
@jstone-lucasfilm I don't know if MrDoob sees the same thing, but as we are developing `WebGPURenderer` we could study a way to port `MaterialX` and share the same algorithms for both WebGL and WebGPU, and I believe we can move forward in implementing a `MeshStandardSurfaceMaterial` from Autodesk for both backends.

**Pull request merged**